### PR TITLE
Fixed #27728 -- Allowed overriding admin templatetag's templates.

### PR DESCRIPTION
**Pull request by rsalmaso**
ticket https://code.djangoproject.com/ticket/27728

### CODE REVIEW

**Comment by collinanderson** [comment]
I would for sure personally use the two object tools overrides, and maybe the result_list too. Do you have use cases for the other overrides? I think there's some advantage to not over-doing it with so many sections.

**Comment by rsalmaso** [comment]
I have a project where I already have overriden all of them on model case. Other than, the ticket references other use cases (some really old!).

File: django/contrib/admin/templates/admin/change_form.html [hunk 1]
```diff
@@ -24,19 +24,7 @@
...
-    <li>
-        {% url opts|admin_urlname:'history' original.pk|admin_urlquote as history_url %}
-        <a href="{% add_preserved_filters history_url %}" class="historylink">{% trans "History" %}</a>
-    </li>
-    {% if has_absolute_url %}<li><a href="{{ absolute_url }}" class="viewsitelink">{% trans "View on site" %}</a></li>{% endif %}
-    {% endblock %}
-  </ul>
-{% endif %}{% endif %}
-{% endblock %}
+{% block object-tools %}{% change_form_object_tools %}{% endblock %}
```
Comment (timgraham) [inline django/contrib/admin/templates/admin/change_form.html]: Is there a need to add another template/template tag for cases liek this when you can already override the template block? I share Collin's concern that adding more levels of indirection makes things more complicated.

Comment (rsalmaso) [reply django/contrib/admin/templates/admin/change_form.html]: Basically I've changed all object-tools with my version on root admin template, and for some model I've added a custom one (essentially an `{% extend "admin/change_{list,form}_object_tools.html" %}`.
I find simpler than copying in my project `/admin/change_{list,form}.html` (cannot extends from django `/admin/change_{list,form}.html` if I recall correctly), as I don't have to copy unnecessary template data. It's only an annoyance, not a real problem.
Can rid both commit if unnecessary.

**Comment by rsalmaso** [comment]
For object-tools I've replaced all object-tools with my version on root admin template, and for some model I've added a custom one (essentially an `{% extend "admin/change_{list,form}_object_tools.html" %}`.
I find simpler than copying in my project `/admin/change_{list,form}.html` because if I recall correctly I cannot extends from django `/admin/change_{list,form}.html` (same level).
For change_list_results.html I use to add some custom elements before and after the list and not copying the list itself.

**Comment by rsalmaso** [comment]
Discussion on django-developers https://groups.google.com/forum/#!topic/django-developers/9_xKJ7GrRhE (currently there is no response, so I think there is no interest in it)

**Comment by rsalmaso** [comment]
rebased and updated to master 2.1

**Comment by rsalmaso** [comment]
Rebased and updated to latest master.
Let me know if there is interest to support and trying to merge this PR or close it.

File: django/contrib/admin/templates/admin/change_form.html (hunk 1 above)
Comment (rsalmaso) [reply django/contrib/admin/templates/admin/change_form.html]: So you are proposing
```html+django
{% block object-tools %}{% if change %}{% if not popup %}
<ul class="object-tools">
{% block object-tools-items %}{% change_form_object_tools %}{% endblock %}
</ul>
{% endif %}{% endif %}{% endblock %}
```
It doesn't break compatibility, but it is only a minor override. I really don't mind about breaking change (admin is a tool, so I think I could live with that), but if you need more customization you can always override the change_X.html templates.
Tomorrow I'll rewrite this part.

Comment (rsalmaso) [reply django/contrib/admin/templates/admin/change_form.html]: Rewritten, now backward compatibility is preserved.

**Comment by rsalmaso** [comment]
Rewritten the object-tools block to preserve backward compatibility.

**Comment by carltongibson** [comment]
OK @rsalmaso Thanks! 

@collinanderson Thanks for the review. Can I ask you to look again when you have the capacity? (Ta)

**Comment by collinanderson** [comment]
@rsalmaso Thanks so much. I tried it out and it worked for me. Thanks!

**Comment by carltongibson** [comment]
Awesome. Thanks for the extra review @collinanderson!

**Comment by rsalmaso** [comment]
Rebased and fixed conflicts.
It seems to error on psycopg2/psycopg2-binary warning output.

**Comment by carltongibson** [comment]
Hi @rsalmaso. I squashed the commits. This seems to have resolved the build issues. (And we'd need to do it anyway.) Can you just cast your eye over and make sure it all looks OK? 

This looks great the me. (The number of tickets it wraps up is super! 🎉)

**Comment by rsalmaso** [comment]
Hi @carltongibson, it seems all ok!
Thanks.

**Comment by collinanderson** [comment]
Sorry, unfortunately it looks like there's another conflict while this has been sitting here. :/

 I think it's totally ready to be merged otherwise. Just needs someone with write access to merge it.

**Comment by timgraham** [comment]
Don't worry about rebasing it, I'll do it when I take a look.

**Comment by rsalmaso** [comment]
thanks!

**Comment by carltongibson** [comment]
Merged in https://github.com/django/django/commit/5cc28dc752c3ae78456bb835c3ba195489fc26d7. Thanks @rsalmaso (and @collinanderson for reviewing) Great addition!

**Comment by timgraham** [comment]
I made https://github.com/django/django/pull/9736 for my review and posed a couple of questions there.

**Pull request closed without merging**